File: NUnit/Version2/UnitTests/RestAPI/FlowControl/SauceLabsFlowController.cs
using System;
using System.Threading;
using RestSharp;
using UnitTests.RestAPI.FlowControl.Base;

namespace UnitTests.RestAPI.FlowControl {
    internal class SauceLabsFlowController : FlowController {
        public override void ControlFlow() {
            while(TooManyTests()) {
                Thread.Sleep(SauceryConstants.SAUCELABS_FLOW_WAIT);
            }
        }

        protected override bool TooManyTests() {
            //int maxParallelMacSessionsAllowed;  //Possible future use.
            var json = GetJsonResponseForUser(SauceryConstants.ACCOUNT_CONCURRENCY_REQUEST);
            var remainingSection = ExtractJsonSegment(json, json.IndexOf("\"remaining", StringComparison.Ordinal), json.Length - 3);
            var flowControl = SimpleJson.DeserializeObject<FlowControl>(remainingSection);
            return flowControl.remaining.overall <= 0;
        }
    }
}
/*
 * Copyright Andrew Gray, Full Circle Solutions
 * Date: 12th July 2014
 * 
 */

File: JUnit/saucery-core/src/main/java/au/com/fullcirclesolutions/saucery/restapi/flowcontrol/SauceLabsFlowController.java
package au.com.fullcirclesolutions.saucery.restapi.flowcontrol;

import au.com.fullcirclesolutions.saucery.restapi.flowcontrol.json.FlowControl;
import au.com.fullcirclesolutions.saucery.utils.SauceryConstants;
import com.fasterxml.jackson.databind.ObjectMapper;

import com.saucelabs.saucerest.SauceREST;
import java.io.IOException;

public class SauceLabsFlowController extends FlowController {
    public String UserName;
    public String AccessKey;
    
    public SauceLabsFlowController(String username, String accessKey) {
        UserName = username;
        AccessKey = accessKey;
    }

    @Override
    public void ControlFlow() {
        while (TooManyTests()) {
            try {
                Thread.sleep(SauceryConstants.SAUCELABS_FLOW_WAIT);
            } catch (InterruptedException e) {
            }
        }
    }

    @Override
    protected boolean TooManyTests() {
        SauceREST sauceRest = new SauceREST(UserName, AccessKey);
        ObjectMapper mapper = new ObjectMapper();
        String remaining = getRemainingSection(sauceRest.getConcurrency());
        
        try {
            FlowControl flowcontrol = mapper.readValue(remaining, FlowControl.class);
            return flowcontrol.remaining.overall <= 0;
        } catch (IOException io){
            return true;
        }
    }
    
    private static String getRemainingSection(String json){
        String remainingSection = json.substring(json.indexOf("\"remaining"), json.length() - 3);
        return String.format("{%s}", remainingSection);
    }
}
/*
 * Copyright: Andrew Gray, Full Circle Solutions
 * Date: 26th March 2014
 */